db/virtual_table.cc

```diff
@@ -76,9 +76,31 @@ mutation_source memtable_filling_virtual_table::as_mutation_source() {
             throw std::runtime_error("streamed_mutation::forwarding::yes unsupported in virtual tables");
         }
 
-        auto populate = [this, mt = make_lw_shared<memtable>(schema()), s, range, permit, slice, pc, trace_state, fwd, fwd_mr] () mutable {
-            return execute(*mt).then([this, mt, s, &range, &permit, &slice, &pc, &trace_state, &fwd, &fwd_mr] () {
-                auto rd = make_restricted_flat_reader(mt->as_data_source(), s, permit, range, slice, pc, trace_state, fwd, fwd_mr);
+        // We cannot pass the partition_range directly to execute()
+        // because it is not guaranteed to be alive until execute() resolves.
+        // It is only guaranteed to be alive as long as the returned reader is alive.
+        // We achieve safety by mediating access through this wrapped. When the reader
+        // dies, pr is cleared and execute() will get an exception.
```

Comment:
Since population is invoked synchronously with reader invocations now, we could simplify this. Assuming you also fix the timeouting to not leave execute() behind. A partition range passed to the reader must be alive around reader invocations, so it will also be alive around execute(). abandon() is also not needed now.

Reply:
so, with proper timeouting, we can get rid of the `query_restrictions` object altogether herer, correct?

Reply:
Yes

Reply:
Hm, one thought tho - we are passing `query_restrictions` to `execute`. Right now, the `execute` that takes `query_restrictions` only forwards the call and invokes `execute` without the restrictions. Can this change in the future?
If so, we'd have to keep `query_restrictions`, but otherwise I'll just switch the `execute` call to the one that only takes a memtable

Reply:
The idea was that the implementor of virtual_table could choose whether to override the execute() variant which takes the restrictions or not, depending on whether he wishes to take them into account or not. It's fine to not take them into account in any case.

Instead of passing query_restrictions, you could pass `partition_range&`.

I wouldn't be so concerned about the future here. We control all the code so can adjust it later if need be.